### ISSUE TITLE
array initialized despite `noinit`?

### ISSUE THREAD

**Issue by Guillaume-Helbecque** [chapel-lang/chapel#23669]
### Summary of Problem
When trying to print a `noinit`'ed array, I was surprised to see that no error occurs, and that the later seems to be initialized.

The code sample in https://chapel-lang.org/docs/technotes/noinit.html (here below) mentions that "Reading one of a `noinit`'ed array elements is a memory error - it will read uninitialized memory". Nevertheless, it doesn't produce any errors either.

### Steps to Reproduce

**Source Code:**

```chapel
config const n = 10;
var D = {1..n};
var A: [D] real = noinit;  // A's elements won't be initialized here...

writeln(A); // but, this prints: 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
```

**Compile command:**
`chpl foo.chpl`

**Execution command:**
`./foo`

**Associated Future Test(s):**

### Configuration Information

- Output of `chpl --version`: 1.32.0
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native *
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: none *
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: `gcc 9.4.0`
- (For Cray systems only) Output of `module list`:

**Comment by bradcray** [comment]
@Guillaume-Helbecque : I don't believe that the intention is that you will necessarily get an error when reading a `noinit`d array; merely that the array's values may contain uninitialized memory, so potentially arbitrary/random values.  For example, here's a little example where I allocate and initialize an array, then free it; then allocate a noinit'd array, which potentially uses that same memory again, inheriting its values (though this is not guaranteed), then frees it; then allocates a third, default-init'd array to show the difference.  [On ATO](https://ato.pxeger.com/run?1=m70sOSOxIDVnwYKlpSVpuhY3r1dzKSiUJRYpOFopRBvq6RkaxCoUpSbmKNiCuAY6CkYgwhhEmIAIUxBhBiLMQYQFiLAEEYYGegax1kDDyosyS1Jz8jSUHBUyi60UlHQUHDWtuWq5YBY5YViUl5-Zl1mCotcJptcJRa8zql4ULc4wLc4gLRDvQX0J8y0A) it shows that the noinit'd array reuses memory / doesn't initialize it:

```chapel
{
  var A: [1..10] real = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0];
  writeln("A is: ", A);
}
{
  var B: [1..10] real = noinit;
  writeln("B is: ", B);
}
{
  var C: [1..10] real;
  writeln("C is: ", C);
}
```


```
A is: 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0
B is: 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0
C is: 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
```

Hopefully my point here is clear which is "One should not rely on 'B' having any speific values in it when created in this way."

**Comment by Guillaume-Helbecque** [comment]
Thank you for the explanation and demo, it's fully clear to me now. I think we can close this "issue", which isn't an issue at all.
